dhcp-relay/dhcp_user_xdp.c

```
@@ -0,0 +1,197 @@
+	printf("\n");
+}
+
+static int xdp_link_detach(int ifindex, __u32 xdp_flags)
+{
+	int err;
+
+	if ((err = bpf_set_link_xdp_fd(ifindex, -1, xdp_flags)) < 0) {
+		fprintf(stderr, "ERR: link set xdp unload failed (err=%d):%s\n",
+			err, strerror(-err));
+		return -1;
+	}
+	return 0;
+}
+
+int xdp_link_attach(int ifindex, __u32 xdp_flags, int prog_fd)
+{
+	int err;
+
+	/* libbpf provide the XDP net_device link-level hook attach helper */
+	err = bpf_set_link_xdp_fd(ifindex, prog_fd, xdp_flags);
+
+	if (err < 0) {
+		fprintf(stderr,
+			"ERR: "
+			"ifindex(%d) link set xdp fd failed (%d): %s\n",
+			ifindex, -err, strerror(-err));
+
+		switch (-err) {
+		case EBUSY:
+		case EEXIST:
+			fprintf(stderr, "Hint: XDP already loaded on device\n");
+			break;
+		case EOPNOTSUPP:
+			fprintf(stderr, "Hint: Native-XDP not supported\n");
+			break;
+		default:
+			break;
+		}
+		return -1;
+	}
+
+	return 0;
+}
+
+/* User program takes two  or three arguments
+ * interface name, relay server IP and prog 
+ * unload flag
+*/
+int main(int argc, char **argv)
+{
+	char filename[256] = "dhcp_kern_xdp.o";
+	int prog_fd, err;
+	int opt;
+
+	__u32 xdp_flags = XDP_FLAGS_DRV_MODE;
+	char dev[IF_NAMESIZE] = "";
+	bool do_unload = 0;
+	struct bpf_map *map = NULL;
+	struct bpf_obj *obj = NULL;
+	int map_fd;
+	int key = 0;
+	char server[15] = "";
+	struct in_addr addr;
+	__u16 ifindex;
+
+	while ((opt = getopt_long(argc, argv, "hui:d:m:", options, NULL)) !=
+	       -1) {
+		switch (opt) {
+		case 'i':
+			strncpy(dev, optarg, IF_NAMESIZE);
+			dev[IF_NAMESIZE - 1] = '\0';
+			ifindex = if_nametoindex(dev);
+			if (ifindex <= 0) {
+				printf("Couldn't find ifname:%s \n", dev);
+				return -EINVAL;
+			}
+			break;
+		case 'd':
+			if (inet_aton(optarg, &addr) == 0) {
+				fprintf(stderr,
+					"Couldn't validate IP address:%s\n",
+					optarg);
+				return -EINVAL;
+			}
+			break;
+		case 'm':
+			if (strcmp(optarg, "skb") == 0) {
+				xdp_flags = XDP_FLAGS_SKB_MODE;
+			} else {
```

I think it's better to fail on an invalid value; so rather `else if (strcmp(optarg, "drv") != 0) { fprintf(stderr, "Invalid mode: %s\n", optarg); return -EINVAL; }`